homeassistant/components/mqtt/__init__.py

```diff
@@ -527,35 +537,52 @@ def stop():
 
     @asyncio.coroutine
     def async_subscribe(self, topic, qos):
-        """Subscribe to a topic.
+        """Set up a subscription to a topic with the provided qos.
 
         This method is a coroutine.
         """
         if not isinstance(topic, str):
-            raise HomeAssistantError("topic need to be a string!")
+            raise HomeAssistantError("topic needs to be a string!")
 
-        with (yield from self._paho_lock):
-            if topic in self.subscribed_topics:
-                return
-            self.wanted_topics[topic] = qos
-            result, mid = yield from self.hass.async_add_job(
-                self._mqttc.subscribe, topic, qos)
-
-            _raise_on_error(result)
-            self.progress[mid] = topic
+        self.wanted_topics[topic].append(qos)
+        yield from self._async_perform_subscription(topic, qos)
 
     @asyncio.coroutine
-    def async_unsubscribe(self, topic):
-        """Unsubscribe from topic.
+    def async_unsubscribe(self, topic, qos):
+        """Unsubscribe from a topic with the specified qos.
 
         This method is a coroutine.
         """
-        self.wanted_topics.pop(topic, None)
-        result, mid = yield from self.hass.async_add_job(
-            self._mqttc.unsubscribe, topic)
+        if not isinstance(topic, str):
+            raise HomeAssistantError("topic needs to be a string!")
+        if topic not in self.wanted_topics:
```

Comment:
Silent failure (not raising when something is wrong) is usually a source of future bugs. The application clearly got into a wrong state or has a bug if we try to unsubscribe a subscription twice.

I know the old code did it too, but I think we should remove it. What do you think?

Reply:
You're completely right - this has annoyed me with other components too so I don't know why I repeated this mistake. It's definitely better to fail fast than to have some unexpected behavior around - so I suppose raising an exception would be good to indicate a failed state.

Anyway, this could also be fixed by introducing a `Subscription` object as you mentioned. Working on that 🛠